numpy/core/src/common/npy_argparse.h

```diff
@@ -0,0 +1,151 @@
+#ifndef _NPY_ARGPARSE_H
+#define _NPY_ARGPARSE_H
+
+#include "Python.h"
+#include "numpy/ndarraytypes.h"
+
+/*
+ * This file defines macros to help with keyword argument parsing.
+ * This solves two issues as of now:
+ *   1. Pythons C-API PyArg_* keyword argument parsers are slow, due to
+ *      not caching the strings they use.
+ *   2. It allows the use of METH_ARGPARSE (and `tp_vectorcall`)
+ *      when available in Python, which removes a large chunk of overhead.
+ *
+ * Internally CPython achieves similar things by using a code generator
+ * argument clinic. NumPy may well decide to use argument clinic or a different
+ * solution in the future.
+ */
+
+NPY_NO_EXPORT int
+PyArray_PythonPyIntFromInt(PyObject *obj, int *value);
+
+
+#define _NPY_MAX_KWARGS 15
+
+typedef struct {
+    int npositional;
+    int nargs;
+    int npositional_only;
+    int nrequired;
+    int nkwarg;
+    // char intent[MAX_KWARGS];
+    /* Null terminated list of keyword argument name strings */
+    PyObject *kw_strings[_NPY_MAX_KWARGS];
+} _NpyArgParserCache;
+
+
+/*
+ * Macros to help support `METH_FASTCALL` when available and fall back to
+ * normal parsing when not.
+ * The signature of function is different in the two cases, so this switches.
+ */
+#ifdef METH_FASTCALL
+
+#if PY_VERSION_HEX >= 0x03070000
+    #define NPY_METH_KWARGS_FASTCALL METH_FASTCALL | METH_KEYWORDS
+#else
+    /* Python 3.6 _officially_ did not support METH_FASTCALL, but it does. */
+    #define NPY_METH_KWARGS_FASTCALL METH_FASTCALL
+#endif
+#define NPY_PARAMS_DEF \
+    PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames
+#define NPY_PARAMS_PASS args, len_args, kwnames
+#define NPY_PARAMS_GET_ARG(i) args[i]
+#define NPY_PARAMS_DEFINE_LEN_ARGS do {} while(0)
+#define NPY_PARAMS_KWARG_OBJ kwnames
+
+#else  /* METH_FASTCALL */
+
+#define NPY_METH_KWARGS_FASTCALL METH_VARARGS | METH_KEYWORDS
+#define NPY_PARAMS_DEF PyObject *args, PyObject *kwargs
+#define NPY_PARAMS_PASS args, kwargs
+#define NPY_PARAMS_GET_ARG(i) PyTuple_GET_ITEM(args, i)
+/* Allow args == NULL, for when all args are already caught by `*args` */
+#define NPY_PARAMS_DEFINE_LEN_ARGS Py_ssize_t len_args = \
+        (args == NULL) ? 0 : PyTuple_GET_SIZE(args)
+#define NPY_PARAMS_KWARG_OBJ kwargs
+
+#endif  /* METH_FASTCALL */
+
+
+#define NPY_PREPARE_ARGPARSER \
+        static _NpyArgParserCache __argparse_cache = {-1};      \
```

Comment:
Not sure I understand why this is static

Reply:
Ah, think I get it now, nevermind

Reply:
To hold the interned strings. I could also do a `static void *cache = NULL;` in each of the functions instead, if that is simpler. There needs to be some place to hold the kwarg string pointers.